### ISSUE TITLE
Secrets are to be kept secret.

### ISSUE THREAD

**Issue by xekoukou** [xekoukou/Agtor#2]
Secrets is the way we define scope. An actor creates a secret and it sends it to other actors he wants to communicate with.

If we have two actors *A* and *B* that communicate through a secret *s*, then the type of the parallel composition A || B should not contain those private communications. In other words, the type system should know how many people the secret has been sent to.

If for example A || B send the secret to a third actor C, then obviously the private communications are no longer private.
Thus after A || B sent the secret, again the type system should reflect that anyone can join the private conversation.

A || B || C is still closed. The way Pi-calculus handles this is to have the *ν_* syntax, together with its structural congruence rule.

ν (A || B) || C would become ν (Α' || Β || C'). But with this structural rule we can't really  know the type of A || B independently of C.

Or to put it differently we are interested in the type of A || B independently of its environment.

As we saw , the change of the scope happens every time we perform a communication of a secret.
If by (D || C) we mean the type that is constructed by the parallel composition of D || C , then this type should take into account the case where we send the secret from D to C, ie internally and when we send the secret to the outside world.
It is at exactly in those two cases that we need to limit the scope and expand it.

In other words the parallel composition of two systems replaces the operator ν_ used in Pi-calculus.

**Comment by xekoukou** [comment]
I have implemented this. 
A. Scope affects the reduction, we have internal reduction and external one, 
B. internal communication of a secret between two process ( a || b ) extends the scope to both, but not the outer world.

C. External communication removes the scope locally. it is up to the greater context to reintroduce it with B.
